File: js/stock/evidencePolicy.js
(function () {
    window.StockPhotonicStock = window.StockPhotonicStock || {};

    const TIERS = {
        verified: {
            key: 'verified',
            label: 'Verified',
            shortLabel: 'VERIFIED',
            graphLabel: 'Verified',
            rank: 4,
            color: '#7cffc8',
            description: 'SEC, official company, or strong source-backed production evidence.'
        },
        strong_inferred: {
            key: 'strong_inferred',
            label: 'Strong inferred',
            shortLabel: 'STRONG',
            graphLabel: 'Strong inferred',
            rank: 3,
            color: '#7dd3fc',
            description: 'Obvious public ecosystem or competitive relationship from stable metadata. Visibility label only, not official proof.'
        },
        context_only: {
            key: 'context_only',
            label: 'Context only',
            shortLabel: 'CONTEXT',
            graphLabel: 'Context only',
            rank: 2,
            color: '#c4b5fd',
            description: 'Research, OpenAlex, topic, overlap, or weak ecosystem context. Not relationship proof.'
        },
        needs_review: {
            key: 'needs_review',
            label: 'Needs review',
            shortLabel: 'REVIEW',
            graphLabel: 'Needs review',
            rank: 1,
            color: '#fb923c',
            description: 'Ambiguous, candidate-only, weak, conflicting, or unresolved relationship signal.'
        }
    };

    const REVIEWER_DECISION_STATES = {
        accepted_for_visibility: {
            key: 'accepted_for_visibility',
            label: 'Accepted for visibility',
            shortLabel: 'VISIBLE',
            description: 'Safe to show in the graph with its tier label; does not authorize production promotion.'
        },
        accepted_for_review: {
            key: 'accepted_for_review',
            label: 'Accepted for review',
            shortLabel: 'REVIEW',
            description: 'Review queue item; no automatic promotion.'
        },
        pending_preview: {
            key: 'pending_preview',
            label: 'Pending preview',
            shortLabel: 'PENDING',
            description: 'Reviewer has not approved preview progression yet.'
        },
        approved_for_preview: {
            key: 'approved_for_preview',
            label: 'Approved for preview',
            shortLabel: 'PREVIEW',
            description: 'Reviewer approved preview visibility only; no production write is authorized.'
        },
        approved_for_promotion_review: {
            key: 'approved_for_promotion_review',
            label: 'Approved for promotion review',
            shortLabel: 'PROMO REVIEW',
            description: 'Reviewer approved deeper manual promotion review; promotion remains explicit and validated.'
        },
        blocked: {
            key: 'blocked',
            label: 'Blocked',
            shortLabel: 'BLOCKED',
            description: 'Cannot move forward without resolving missing or conflicting inputs.'
        },
        weak_signal: {
            key: 'weak_signal',
            label: 'Weak signal',
            shortLabel: 'WEAK',
            description: 'Visible only as weak context until stronger evidence is attached.'
        },
        enrichment_only: {
            key: 'enrichment_only',
            label: 'Enrichment only',
            shortLabel: 'ENRICH',
            description: 'Can guide research, but cannot prove or promote a relationship.'
        },
        ready_for_promotion_review: {
            key: 'ready_for_promotion_review',
            label: 'Ready for promotion review',
            shortLabel: 'PROMO REVIEW',
            description: 'Candidate has stronger inputs but still requires manual preview, promotion, and validation.'
        },
        production_candidate: {
            key: 'production_candidate',
            label: 'Production candidate',
            shortLabel: 'PROD CAND',
            description: 'Reviewer marked the record for manual promotion tooling and validation; no automatic write is authorized.'
        },
        deferred: {
            key: 'deferred',
            label: 'Deferred',
            shortLabel: 'DEFER',
            description: 'Reviewer postponed the record without promotion authority.'
        }
    };

    const TRUSTED_RELATIONSHIP_CLASSES = {
        competitor: {
            key: 'competitor',
            label: 'Competitor',
            shortLabel: 'COMP',
            safeFastTrack: true,
            minimumConfidence: 3,
            minimumStrength: 0.62,
            keywords: ['competitor', 'competition', 'competes', 'peer', 'rival', 'market structure', 'direct peer']
        },
        ecosystem_overlap: {
            key: 'ecosystem_overlap',
            label: 'Ecosystem overlap',
            shortLabel: 'ECO',
            safeFastTrack: true,
            minimumConfidence: 4,
            minimumStrength: 0.72,
            keywords: ['ecosystem', 'overlap', 'market structure', 'same market', 'shared market', 'platform exposure']
        },
        supplier_ecosystem: {
            key: 'supplier_ecosystem',
            label: 'Supplier ecosystem',
            shortLabel: 'SUP ECO',
            safeFastTrack: false,
            minimumConfidence: 4,
            minimumStrength: 0.78,
            keywords: ['supplier', 'supply', 'customer', 'vendor', 'dependency', 'supply chain']
        },
        cloud_hyperscaler_exposure: {
            key: 'cloud_hyperscaler_exposure',
            label: 'Cloud / hyperscaler exposure',
            shortLabel: 'CLOUD',
            safeFastTrack: true,
            minimumConfidence: 4,
            minimumStrength: 0.70,
            keywords: ['aws', 'azure', 'google cloud', 'gcp', 'oci', 'oracle cloud', 'cloud', 'hyperscaler', 'data center', 'cloud ai']
        },
        semiconductor_supply_chain: {
            key: 'semiconductor_supply_chain',
            label: 'Semiconductor supply chain',
            shortLabel: 'SEMI',
            safeFastTrack: true,
            minimumConfidence: 4,
            minimumStrength: 0.70,
            keywords: ['semiconductor', 'foundry', 'hbm', 'memory', 'lithography', 'wafer', 'fab', 'accelerator', 'gpu', 'chip']
        },
        financial_infrastructure_overlap: {
            key: 'financial_infrastructure_overlap',
            label: 'Financial infrastructure overlap',
            shortLabel: 'FIN',
            safeFastTrack: true,
            minimumConfidence: 4,
            minimumStrength: 0.68,
            keywords: ['payment', 'payments', 'card', 'issuer', 'acquirer', 'bank', 'exchange', 'asset manager', 'capital markets', 'financial infrastructure']
        }
    };

    const TRUSTED_CLASS_LIST = Object.values(TRUSTED_RELATIONSHIP_CLASSES);

    function normalizeText(value) {
        return String(value || '').trim().toLowerCase();
    }

    function normalizeKey(value) {
        return normalizeText(value).replace(/[\s-]+/g, '_');
    }

    function hasAnyKeyword(text, keywords) {
        return keywords.some(keyword => text.includes(keyword));
    }

    function getValidSourceUrls(record) {
        const candidate = record?.candidate || {};
        const values = [
            ...(Array.isArray(record?.source_urls) ? record.source_urls : []),
            ...(Array.isArray(candidate.source_urls) ? candidate.source_urls : []),
            record?.source_url,
            record?.filing_url,
            record?.sec_url,
            record?.url,
            candidate.archive_url,
            candidate.source_url,
            candidate.filing_url,
            candidate.sec_url
        ];
        return [...new Set(values.map(url => String(url || '').trim()).filter(url => /^https?:\/\//i.test(url)))];
    }

    function getSourceHost(url) {
        try {
            return new URL(url).hostname.replace(/^www\./i, '').toLowerCase();
        } catch (error) {
            return '';
        }
    }

    function getSourcePath(url) {
        try {
            const parsed = new URL(url);
            return `${parsed.hostname}${parsed.pathname}`.toLowerCase();
        } catch (error) {
            return String(url || '').toLowerCase();
        }
    }

    function getConfidenceScore(record) {
        const raw = Number(record?.confidence_score ?? record?.confidence ?? record?.confidence_hint ?? record?.candidate?.confidence_hint);
        if (!Number.isFinite(raw) || raw <= 0) return null;
        return raw <= 1 ? Math.round(raw * 5) : Math.max(1, Math.min(5, Math.round(raw)));
    }

    function getStrengthScore(record) {
        const raw = Number(record?.strength);
        return Number.isFinite(raw) ? Math.max(0, Math.min(1, raw)) : 0;
    }

    function isCandidateRecord(record) {
        const state = normalizeText(record?.source_status || record?.review_status || record?.candidate?.review_status);
        return Boolean(
            record?.isSecPreviewLink ||
            record?.is_candidate_preview ||
            record?.candidate ||
            state.includes('candidate') ||
            state.includes('pending_review')
        );
    }

    function isOpenAlexContext(record) {
        const tags = Array.isArray(record?.evidence_tags) ? record.evidence_tags.join(' ') : '';
        const text = [
            record?.source,
            record?.source_type,
            record?.provider,
            record?.origin,
            record?.provenance,
            record?.relationship_type,
            record?.type,
            tags
        ].map(normalizeText).join(' ');
        return text.includes('openalex');
    }

    function isSecBacked(record) {
        const text = getRelationshipText(record);
        return text.includes('sec filing') || getValidSourceUrls(record).some(url => /(\.|^)sec\.gov$/i.test(getSourceHost(url)));
    }

    function hasOfficialSource(record) {
        const sourceReview = window.StockPhotonicStock?.sourceReview;
        const categories = sourceReview?.getSourceHostDiversity?.(record)?.categoryKeys || [];
        if (categories.some(key => ['sec_source', 'official_company_ir', 'official_partner_customer_page'].includes(key))) {
            return true;
        }

        return getValidSourceUrls(record).some(url => {
            const host = getSourceHost(url);
            const path = getSourcePath(url);
            return /(\.|^)sec\.gov$/i.test(host) ||
                /(investor|investors|ir\.|\/ir\/|shareholder|sec-filings|annual-report|quarterly-results|news-releases|partner|partners|customer|customers|case-study|press|collaboration|alliance)/i.test(path);
        });
    }

    function getEvidenceCount(record) {
        const sourceCount = getValidSourceUrls(record).length;
        const snippetCount = record?.evidence_snippet || record?.candidate?.evidence_snippet ? 1 : 0;
        const sourceLabelCount = record?.source_label || record?.provenance ? 1 : 0;
        return sourceCount + snippetCount + sourceLabelCount;
    }

    function getEndpointText(endpoint) {
        if (!endpoint || typeof endpoint !== 'object') return '';
        return [
            endpoint.ticker,
            endpoint.name,
            endpoint.sector,
            endpoint.industry,
            endpoint.industryGroup
        ].map(normalizeText).filter(Boolean).join(' ');
    }

    function getRelationshipText(record) {
        const candidate = record?.candidate || {};
        return [
            record?.relationship_type,
            record?.type,
            record?.raw_type,
            record?.label,
            record?.relationship_summary,
            record?.provenance,
            record?.source_label,
            record?.evidence_snippet,
            candidate.relationship_type,
            candidate.evidence_snippet,
            candidate.source,
            candidate.review_status,
            getEndpointText(record?.source),
            getEndpointText(record?.target)
        ].map(normalizeText).filter(Boolean).join(' ');
    }

    function getTrustedRelationshipClass(record) {
        const text = getRelationshipText(record);
        const rawType = normalizeKey(record?.relationship_type || record?.type || record?.candidate?.relationship_type);
        const endpointText = `${getEndpointText(record?.source)} ${getEndpointText(record?.target)}`;

        if (rawType.includes('competitor') || rawType.includes('competition') || rawType.includes('peer') || hasAnyKeyword(text, TRUSTED_RELATIONSHIP_CLASSES.competitor.keywords)) {
            return buildClassResult(TRUSTED_RELATIONSHIP_CLASSES.competitor, 'relationship type or label indicates peer competition');
        }
        if (rawType.includes('hyperscaler') || rawType.includes('cloud') || hasAnyKeyword(text, TRUSTED_RELATIONSHIP_CLASSES.cloud_hyperscaler_exposure.keywords) || endpointText.includes('cloud / big tech')) {
            return buildClassResult(TRUSTED_RELATIONSHIP_CLASSES.cloud_hyperscaler_exposure, 'metadata fits cloud or hyperscaler ecosystem exposure');
        }
        if (rawType.includes('semiconductor') || hasAnyKeyword(text, TRUSTED_RELATIONSHIP_CLASSES.semiconductor_supply_chain.keywords) || endpointText.includes('ai / semiconductors')) {
            return buildClassResult(TRUSTED_RELATIONSHIP_CLASSES.semiconductor_supply_chain, 'metadata fits semiconductor ecosystem context');
        }
        if (rawType.includes('ownership') || rawType.includes('capital') || rawType.includes('financial') || hasAnyKeyword(text, TRUSTED_RELATIONSHIP_CLASSES.financial_infrastructure_overlap.keywords) || endpointText.includes('payments / financial infrastructure')) {
            return buildClassResult(TRUSTED_RELATIONSHIP_CLASSES.financial_infrastructure_overlap, 'metadata fits financial infrastructure or payment-network overlap');
        }
        if (rawType.includes('supplier') || rawType.includes('customer') || rawType === 'supply' || hasAnyKeyword(text, TRUSTED_RELATIONSHIP_CLASSES.supplier_ecosystem.keywords)) {
            return buildClassResult(TRUSTED_RELATIONSHIP_CLASSES.supplier_ecosystem, 'metadata indicates supplier or customer ecosystem context');
        }
        if (rawType.includes('ecosystem') || hasAnyKeyword(text, TRUSTED_RELATIONSHIP_CLASSES.ecosystem_overlap.keywords)) {
            return buildClassResult(TRUSTED_RELATIONSHIP_CLASSES.ecosystem_overlap, 'metadata indicates ecosystem overlap');
        }
        return null;
    }

    function buildClassResult(definition, reason) {
        return {
            ...definition,
            reason
        };
    }

    function isStrongInferredEligible(record, trustedClass) {
        if (!trustedClass?.safeFastTrack || isCandidateRecord(record) || isOpenAlexContext(record)) return false;
        if (hasOfficialSource(record)) return false;

        const confidence = getConfidenceScore(record) || 0;
        const strength = getStrengthScore(record);
        return confidence >= trustedClass.minimumConfidence || strength >= trustedClass.minimumStrength;
    }

    function hasUnresolvedCandidateEndpoint(record) {
        if (!isCandidateRecord(record)) return false;
        const candidate = record?.candidate || record || {};
        return !String(candidate.source_ticker || record?.sourceTicker || '').trim() ||
            !String(candidate.target_ticker || record?.targetTicker || '').trim();
    }

    function getEvidencePolicy(record, options = {}) {
        const trustedClass = getTrustedRelationshipClass(record);
        const confidenceScore = getConfidenceScore(record);
        const sourceUrls = getValidSourceUrls(record);
        const evidenceCount = Number(options.evidenceCount ?? record?.evidence_count ?? getEvidenceCount(record)) || 0;
        const candidate = isCandidateRecord(record);
        const openAlex = isOpenAlexContext(record);
        const official = hasOfficialSource(record);
        const secBacked = isSecBacked(record);
        const strongInferred = isStrongInferredEligible(record, trustedClass);
        const hasSource = sourceUrls.length > 0;
        const unresolved = hasUnresolvedCandidateEndpoint(record);

        let tier = TIERS.needs_review;
        let decisionState = REVIEWER_DECISION_STATES.accepted_for_review;
        let explanation = 'Review required before this relationship can be treated as production-quality evidence.';

        if (openAlex) {
            tier = TIERS.context_only;
            decisionState = REVIEWER_DECISION_STATES.enrichment_only;
            explanation = 'Context-only OpenAlex enrichment. It can guide research but is not relationship proof.';
        } else if (candidate) {
            tier = TIERS.needs_review;
            decisionState = unresolved
                ? REVIEWER_DECISION_STATES.blocked
                : secBacked && hasSource && evidenceCount > 0 && (confidenceScore || 0) >= 4
                    ? REVIEWER_DECISION_STATES.ready_for_promotion_review
                    : REVIEWER_DECISION_STATES.accepted_for_review;
            explanation = decisionState.key === 'ready_for_promotion_review'
                ? 'SEC-backed candidate evidence is ready for manual promotion review; promotion is still manual.'
                : 'Candidate or preview evidence requires manual review before any production promotion.';
        } else if (official || secBacked || (hasSource && evidenceCount > 0 && (confidenceScore || 0) >= 4)) {
            tier = TIERS.verified;
            decisionState = REVIEWER_DECISION_STATES.accepted_for_visibility;
            explanation = secBacked
                ? 'Verified SEC-backed production relationship.'
                : 'Verified source-backed production relationship.';
        } else if (strongInferred) {
            tier = TIERS.strong_inferred;
            decisionState = REVIEWER_DECISION_STATES.accepted_for_visibility;
            explanation = `Strong inferred ${trustedClass.label.toLowerCase()} relationship. Safe for graph visibility, not official partnership proof.`;
        } else if (trustedClass || hasSource || evidenceCount > 0) {
            tier = TIERS.context_only;
            decisionState = trustedClass?.safeFastTrack
                ? REVIEWER_DECISION_STATES.weak_signal
                : REVIEWER_DECISION_STATES.accepted_for_review;
            explanation = trustedClass
                ? `Context-only ${trustedClass.label.toLowerCase()} signal; attach stronger sources before promotion claims.`
                : 'Context-only evidence signal; not enough for verified relationship status.';
        }

        return {
            tier,
            tierKey: tier.key,
            trustedClass,
            trustedClassKey: trustedClass?.key || '',
            trustedClassLabel: trustedClass?.label || '',
            reviewerDecisionState: decisionState,
            reviewerDecisionKey: decisionState.key,
            explanation,
            confidenceScore,
            evidenceCount,
            sourceCount: sourceUrls.length,
            officialSource: official,
            secBacked,
            openAlex,
            candidate,
            fastTrackVisibility: tier.key === TIERS.strong_inferred.key,
            manualPromotionAllowed: false,
            sourceUrls
        };
    }

    function getEvidenceTier(record, options = {}) {
        return getEvidencePolicy(record, options).tier;
    }

    function getReviewerDecisionState(record, options = {}) {
        return getEvidencePolicy(record, options).reviewerDecisionState;
    }

    function shouldReduceReviewPressure(record, options = {}) {
        const policy = getEvidencePolicy(record, options);
        return policy.tier.key === TIERS.strong_inferred.key &&
            policy.reviewerDecisionState.key === REVIEWER_DECISION_STATES.accepted_for_visibility.key;
    }

    function getEvidencePolicyTags(record, options = {}) {
        const policy = getEvidencePolicy(record, options);
        const tags = [
            policy.tier.shortLabel,
            policy.trustedClass?.shortLabel || '',
            policy.reviewerDecisionState.shortLabel
        ];
        if (policy.fastTrackVisibility) tags.push('FAST-TRACK VISIBLE');
        if (policy.openAlex) tags.push('OPENALEX CONTEXT');
        if (policy.candidate) tags.push('REVIEW ONLY');
        return [...new Set(tags.filter(Boolean))];
    }

    function summarizeEvidencePolicies(records) {
        const tierCounts = new Map();
        const trustedClassCounts = new Map();
        let fastTrackVisibilityCount = 0;
        let reviewRequiredCount = 0;

        (records || []).forEach(record => {
            const policy = getEvidencePolicy(record);
            tierCounts.set(policy.tier.key, (tierCounts.get(policy.tier.key) || 0) + 1);
            if (policy.trustedClassKey) {
                trustedClassCounts.set(policy.trustedClassKey, (trustedClassCounts.get(policy.trustedClassKey) || 0) + 1);
            }
            if (policy.fastTrackVisibility) fastTrackVisibilityCount += 1;
            if (policy.tier.key === TIERS.needs_review.key) reviewRequiredCount += 1;
        });

        return {
            total: (records || []).length,
            tierCounts: [...tierCounts.entries()].sort((a, b) => (TIERS[b[0]]?.rank || 0) - (TIERS[a[0]]?.rank || 0)),
            trustedClassCounts: [...trustedClassCounts.entries()].sort((a, b) => b[1] - a[1] || a[0].localeCompare(b[0])),
            fastTrackVisibilityCount,
            reviewRequiredCount
        };
    }

    window.StockPhotonicStock.evidencePolicy = {
        TIERS,
        TIER_LIST: Object.values(TIERS).sort((a, b) => b.rank - a.rank),
        REVIEWER_DECISION_STATES,
        REVIEWER_DECISION_STATE_LIST: Object.values(REVIEWER_DECISION_STATES),
        TRUSTED_RELATIONSHIP_CLASSES,
        TRUSTED_CLASS_LIST,
        normalizeText,
        getValidSourceUrls,
        getConfidenceScore,
        getStrengthScore,
        getEvidenceCount,
        isCandidateRecord,
        isOpenAlexContext,
        isSecBacked,
        hasOfficialSource,
        getRelationshipText,
        getTrustedRelationshipClass,
        getEvidencePolicy,
        getEvidenceTier,
        getReviewerDecisionState,
        shouldReduceReviewPressure,
        getEvidencePolicyTags,
        summarizeEvidencePolicies
    };
})();

File: scripts/evidence_policy.py
#!/usr/bin/env python3
"""Shared review-only evidence policy helpers for StockPhotonic scripts."""

from __future__ import annotations

from collections import Counter
from typing import Any
from urllib.parse import urlparse


TIERS: dict[str, dict[str, Any]] = {
    "verified": {
        "key": "verified",
        "label": "Verified",
        "short_label": "VERIFIED",
        "rank": 4,
    },
    "strong_inferred": {
        "key": "strong_inferred",
        "label": "Strong inferred",
        "short_label": "STRONG",
        "rank": 3,
    },
    "context_only": {
        "key": "context_only",
        "label": "Context only",
        "short_label": "CONTEXT",
        "rank": 2,
    },
    "needs_review": {
        "key": "needs_review",
        "label": "Needs review",
        "short_label": "REVIEW",
        "rank": 1,
    },
}

REVIEWER_DECISION_STATES: dict[str, dict[str, Any]] = {
    "accepted_for_visibility": {
        "key": "accepted_for_visibility",
        "label": "Accepted for visibility",
        "short_label": "VISIBLE",
    },
    "accepted_for_review": {
        "key": "accepted_for_review",
        "label": "Accepted for review",
        "short_label": "REVIEW",
    },
    "blocked": {
        "key": "blocked",
        "label": "Blocked",
        "short_label": "BLOCKED",
    },
    "weak_signal": {
        "key": "weak_signal",
        "label": "Weak signal",
        "short_label": "WEAK",
    },
    "enrichment_only": {
        "key": "enrichment_only",
        "label": "Enrichment only",
        "short_label": "ENRICH",
    },
    "ready_for_promotion_review": {
        "key": "ready_for_promotion_review",
        "label": "Ready for promotion review",
        "short_label": "PROMO REVIEW",
    },
}

TRUSTED_CLASSES: dict[str, dict[str, Any]] = {
    "competitor": {
        "key": "competitor",
        "label": "Competitor",
        "short_label": "COMP",
        "safe_fast_track": True,
        "minimum_confidence": 3,
        "minimum_strength": 0.62,
        "keywords": ("competitor", "competition", "competes", "peer", "rival", "market structure"),
    },
    "ecosystem_overlap": {
        "key": "ecosystem_overlap",
        "label": "Ecosystem overlap",
        "short_label": "ECO",
        "safe_fast_track": True,
        "minimum_confidence": 4,
        "minimum_strength": 0.72,
        "keywords": ("ecosystem", "overlap", "same market", "shared market", "platform exposure"),
    },
    "supplier_ecosystem": {
        "key": "supplier_ecosystem",
        "label": "Supplier ecosystem",
        "short_label": "SUP ECO",
        "safe_fast_track": False,
        "minimum_confidence": 4,
        "minimum_strength": 0.78,
        "keywords": ("supplier", "supply", "customer", "vendor", "dependency", "supply chain"),
    },
    "cloud_hyperscaler_exposure": {
        "key": "cloud_hyperscaler_exposure",
        "label": "Cloud / hyperscaler exposure",
        "short_label": "CLOUD",
        "safe_fast_track": True,
        "minimum_confidence": 4,
        "minimum_strength": 0.70,
        "keywords": ("aws", "azure", "google cloud", "gcp", "oci", "cloud", "hyperscaler", "data center"),
    },
    "semiconductor_supply_chain": {
        "key": "semiconductor_supply_chain",
        "label": "Semiconductor supply chain",
        "short_label": "SEMI",
        "safe_fast_track": True,
        "minimum_confidence": 4,
        "minimum_strength": 0.70,
        "keywords": ("semiconductor", "foundry", "hbm", "memory", "lithography", "wafer", "fab", "gpu", "chip"),
    },
    "financial_infrastructure_overlap": {
        "key": "financial_infrastructure_overlap",
        "label": "Financial infrastructure overlap",
        "short_label": "FIN",
        "safe_fast_track": True,
        "minimum_confidence": 4,
        "minimum_strength": 0.68,
        "keywords": ("payment", "payments", "card", "issuer", "acquirer", "bank", "exchange", "asset manager", "financial infrastructure"),
    },
}


def clean_string(value: Any) -> str:
    return str(value or "").strip()


def normalize_text(value: Any) -> str:
    return clean_string(value).lower()


def normalize_key(value: Any) -> str:
    return normalize_text(value).replace("-", "_").replace(" ", "_")


def source_urls(record: dict[str, Any]) -> list[str]:
    candidate = record.get("candidate") if isinstance(record.get("candidate"), dict) else {}
    raw_values: list[Any] = []
    for source in (record, candidate):
        urls = source.get("source_urls")
        if isinstance(urls, list):
            raw_values.extend(urls)
        for key in ("archive_url", "source_url", "filing_url", "sec_url", "url"):
            raw_values.append(source.get(key))
    return sorted(
        {
            clean_string(value)
            for value in raw_values
            if clean_string(value).lower().startswith(("http://", "https://"))
        }
    )


def source_host(url: str) -> str:
    parsed = urlparse(url)
    return parsed.hostname.lower().removeprefix("www.") if parsed.hostname else ""


def source_path(url: str) -> str:
    parsed = urlparse(url)
    return f"{parsed.hostname or ''}{parsed.path or ''}".lower()


def confidence_score(record: dict[str, Any]) -> int | None:
    for key in ("confidence_score", "confidence", "confidence_hint"):
        value = record.get(key)
        if isinstance(value, bool):
            continue
        if isinstance(value, (int, float)) and value > 0:
            return round(value * 5) if value <= 1 else max(1, min(5, round(value)))
    candidate = record.get("candidate") if isinstance(record.get("candidate"), dict) else {}
    value = candidate.get("confidence_hint")
    if isinstance(value, (int, float)) and not isinstance(value, bool) and value > 0:
        return round(value * 5) if value <= 1 else max(1, min(5, round(value)))
    return None


def strength_score(record: dict[str, Any]) -> float:
    value = record.get("strength")
    return max(0.0, min(1.0, float(value))) if isinstance(value, (int, float)) and not isinstance(value, bool) else 0.0


def endpoint_text(company: dict[str, Any] | None) -> str:
    if not isinstance(company, dict):
        return ""
    return " ".join(
        normalize_text(company.get(key))
        for key in ("ticker", "name", "sector", "industry", "industryGroup", "industry_group")
    )


def relationship_text(
    record: dict[str, Any],
    source_company: dict[str, Any] | None = None,
    target_company: dict[str, Any] | None = None,
) -> str:
    candidate = record.get("candidate") if isinstance(record.get("candidate"), dict) else {}
    values = [
        record.get("relationship_type"),
        record.get("type"),
        record.get("raw_type"),
        record.get("label"),
        record.get("relationship_summary"),
        record.get("provenance"),
        record.get("source_label"),
        record.get("evidence_snippet"),
        record.get("source_type"),
        candidate.get("relationship_type"),
        candidate.get("evidence_snippet"),
        candidate.get("source"),
        candidate.get("review_status"),
        endpoint_text(source_company),
        endpoint_text(target_company),
    ]
    return " ".join(normalize_text(value) for value in values if normalize_text(value))


def is_candidate_record(record: dict[str, Any]) -> bool:
    status = normalize_text(record.get("source_status") or record.get("review_status"))
    candidate = record.get("candidate")
    return bool(
        record.get("isSecPreviewLink")
        or record.get("is_candidate_preview")
        or isinstance(candidate, dict)
        or status.startswith("pending")
        or "candidate" in status
    )


def is_openalex_context(record: dict[str, Any]) -> bool:
    values = [
        record.get("source"),
        record.get("source_type"),
        record.get("provider"),
        record.get("origin"),
        record.get("provenance"),
        record.get("relationship_type"),
        record.get("type"),
    ]
    tags = record.get("evidence_tags")
    if isinstance(tags, list):
        values.extend(tags)
    return "openalex" in " ".join(normalize_text(value) for value in values)


def is_sec_backed(record: dict[str, Any]) -> bool:
    text = relationship_text(record)
    return "sec filing" in text or any("sec.gov" in source_host(url) for url in source_urls(record))


def has_official_source(record: dict[str, Any]) -> bool:
    for url in source_urls(record):
        host = source_host(url)
        path = source_path(url)
        if host.endswith("sec.gov"):
            return True
        if any(
            token in path
            for token in (
                "investor",
                "investors",
                "/ir/",
                "shareholder",
                "sec-filings",
                "annual-report",
                "quarterly-results",
                "news-releases",
                "partner",
                "customer",
                "case-study",
                "press",
                "collaboration",
                "alliance",
            )
        ):
            return True
    return False


def evidence_count(record: dict[str, Any]) -> int:
    candidate = record.get("candidate") if isinstance(record.get("candidate"), dict) else {}
    return (
        len(source_urls(record))
        + int(bool(clean_string(record.get("evidence_snippet")) or clean_string(candidate.get("evidence_snippet"))))
        + int(bool(clean_string(record.get("source_label")) or clean_string(record.get("provenance"))))
    )


def trusted_relationship_class(
    record: dict[str, Any],
    source_company: dict[str, Any] | None = None,
    target_company: dict[str, Any] | None = None,
) -> dict[str, Any] | None:
    text = relationship_text(record, source_company, target_company)
    raw_type = normalize_key(record.get("relationship_type") or record.get("type"))
    endpoint_context = f"{endpoint_text(source_company)} {endpoint_text(target_company)}"

    ordered = [
        ("competitor", ("competitor", "competition", "peer")),
        ("cloud_hyperscaler_exposure", ("hyperscaler", "cloud")),
        ("semiconductor_supply_chain", ("semiconductor",)),
        ("financial_infrastructure_overlap", ("ownership", "capital", "financial")),
        ("supplier_ecosystem", ("supplier", "customer", "supply")),
        ("ecosystem_overlap", ("ecosystem",)),
    ]
    for key, raw_tokens in ordered:
        definition = TRUSTED_CLASSES[key]
        if (
            any(token in raw_type for token in raw_tokens)
            or any(keyword in text for keyword in definition["keywords"])
            or (key == "semiconductor_supply_chain" and "ai / semiconductors" in endpoint_context)
            or (key == "cloud_hyperscaler_exposure" and "cloud / big tech" in endpoint_context)
            or (key == "financial_infrastructure_overlap" and "payments / financial infrastructure" in endpoint_context)
        ):
            return {
                **definition,
                "reason": f"metadata fits {definition['label'].lower()}",
            }
    return None


def strong_inferred_eligible(
    record: dict[str, Any],
    trusted_class: dict[str, Any] | None,
) -> bool:
    if not trusted_class or not trusted_class.get("safe_fast_track"):
        return False
    if is_candidate_record(record) or is_openalex_context(record) or has_official_source(record):
        return False
    score = confidence_score(record) or 0
    strength = strength_score(record)
    return score >= int(trusted_class["minimum_confidence"]) or strength >= float(trusted_class["minimum_strength"])


def build_edge_policy(
    record: dict[str, Any],
    source_company: dict[str, Any] | None = None,
    target_company: dict[str, Any] | None = None,
) -> dict[str, Any]:
    trusted_class = trusted_relationship_class(record, source_company, target_company)
    score = confidence_score(record)
    count = evidence_count(record)
    urls = source_urls(record)
    candidate = is_candidate_record(record)
    openalex = is_openalex_context(record)
    official = has_official_source(record)
    sec = is_sec_backed(record)
    strong = strong_inferred_eligible(record, trusted_class)

    tier = TIERS["needs_review"]
    decision = REVIEWER_DECISION_STATES["accepted_for_review"]
    explanation = "Review required before this relationship can be treated as production-quality evidence."

    if openalex:
        tier = TIERS["context_only"]
        decision = REVIEWER_DECISION_STATES["enrichment_only"]
        explanation = "Context-only OpenAlex enrichment; not relationship proof."
    elif candidate:
        tier = TIERS["needs_review"]
        ready = sec and bool(urls) and count > 0 and (score or 0) >= 4
        decision = REVIEWER_DECISION_STATES["ready_for_promotion_review" if ready else "accepted_for_review"]
        explanation = (
            "SEC-backed candidate evidence is ready for manual promotion review; promotion is still manual."
            if ready
            else "Candidate or preview evidence requires manual review before any production promotion."
        )
    elif official or sec or (urls and count > 0 and (score or 0) >= 4):
        tier = TIERS["verified"]
        decision = REVIEWER_DECISION_STATES["accepted_for_visibility"]
        explanation = "Verified SEC-backed production relationship." if sec else "Verified source-backed production relationship."
    elif strong:
        tier = TIERS["strong_inferred"]
        decision = REVIEWER_DECISION_STATES["accepted_for_visibility"]
        explanation = (
            f"Strong inferred {trusted_class['label'].lower()} relationship. "
            "Safe for graph visibility, not official partnership proof."
        )
    elif trusted_class or urls or count > 0:
        tier = TIERS["context_only"]
        decision = REVIEWER_DECISION_STATES["weak_signal" if trusted_class and trusted_class.get("safe_fast_track") else "accepted_for_review"]
        explanation = (
            f"Context-only {trusted_class['label'].lower()} signal; attach stronger sources before promotion claims."
            if trusted_class
            else "Context-only evidence signal; not enough for verified relationship status."
        )

    return {
        "evidence_tier": tier["key"],
        "evidence_tier_label": tier["label"],
        "evidence_tier_short_label": tier["short_label"],
        "evidence_tier_rank": tier["rank"],
        "trusted_relationship_class": trusted_class["key"] if trusted_class else None,
        "trusted_relationship_class_label": trusted_class["label"] if trusted_class else None,
        "trusted_relationship_class_reason": trusted_class.get("reason") if trusted_class else None,
        "trusted_relationship_fast_track": bool(trusted_class and trusted_class.get("safe_fast_track")),
        "reviewer_decision_state": decision["key"],
        "reviewer_decision_label": decision["label"],
        "fast_track_visibility": tier["key"] == "strong_inferred",
        "manual_promotion_allowed": False,
        "confidence_score": score,
        "source_count": len(urls),
        "evidence_count": count,
        "official_source": official,
        "sec_backed": sec,
        "explanation": explanation,
        "review_only": True,
    }


def source_search_query(
    row: dict[str, Any],
    policy: dict[str, Any],
    source_company: dict[str, Any] | None = None,
    target_company: dict[str, Any] | None = None,
) -> str:
    source = clean_string(source_company.get("ticker") if source_company else row.get("source_ticker"))
    target = clean_string(target_company.get("ticker") if target_company else row.get("target_ticker"))
    trusted = clean_string(policy.get("trusted_relationship_class_label"))
    return " ".join(part for part in (source, target, trusted, "company source relationship") if part)


def summarize_policies(policies: list[dict[str, Any]]) -> dict[str, Any]:
    tier_counts = Counter(policy.get("evidence_tier") or "needs_review" for policy in policies)
    class_counts = Counter(
        policy.get("trusted_relationship_class")
        for policy in policies
        if policy.get("trusted_relationship_class")
    )
    return {
        "tier_counts": dict(sorted(tier_counts.items())),
        "trusted_relationship_class_counts": dict(sorted(class_counts.items())),
        "fast_track_visibility_count": sum(1 for policy in policies if policy.get("fast_track_visibility")),
        "needs_review_count": tier_counts.get("needs_review", 0),
        "context_only_count": tier_counts.get("context_only", 0),
        "verified_count": tier_counts.get("verified", 0),
        "strong_inferred_count": tier_counts.get("strong_inferred", 0),
        "review_only": True,
    }
